extensions/azurecore/src/account-provider/azureAccountProvider.ts

```diff
@@ -3,446 +3,155 @@
-				.then(
-					() => {
-						return account;
-					},
-					() => {
-						account.isStale = true;
-						return account;
-					}
-				);
-			rehydrationTasks.push(task);
+	private getAuthMethod(account?: azdata.Account): AzureAuth {
+		if (this.authMappings.size === 1) {
+			return this.authMappings.values().next().value;
 		}
 
-		// Collect the rehydration tasks and mark the provider as initialized
-		return Promise.all(rehydrationTasks)
-			.then(accounts => {
-				self._isInitialized = true;
-				return accounts;
-			});
-	}
-
-	public prompt(): Thenable<AzureAccount | azdata.PromptFailedResult> {
-		return this.doIfInitialized(() => this.signIn(true));
+		const authType: AzureAuthType = account?.properties?.azureAuthType;
+		if (authType) {
+			return this.authMappings.get(authType);
+		} else {
+			return this.authMappings.get(AzureAuthType.AuthCodeGrant);
+		}
 	}
 
-	public refresh(account: AzureAccount): Thenable<AzureAccount | azdata.PromptFailedResult> {
-		return this.doIfInitialized(() => this.signIn(false));
+	initialize(storedAccounts: azdata.Account[]): Thenable<azdata.Account[]> {
+		return this._initialize(storedAccounts);
 	}
 
-	// PRIVATE METHODS /////////////////////////////////////////////////////
-	private cancelAutoOAuth(): Promise<void> {
-		let self = this;
-
-		if (!this._inProgressAutoOAuth) {
-			console.warn('Attempted to cancel auto OAuth when auto OAuth is not in progress!');
-			return Promise.resolve();
-		}
-
-		// Indicate oauth was cancelled by the user
-		let inProgress = self._inProgressAutoOAuth;
-		self._autoOAuthCancelled = true;
-		self._inProgressAutoOAuth = null;
-
-		// Use the auth context that was originally used to open the polling request, and cancel the polling
-		let context = inProgress.context;
-		context.cancelRequestToGetTokenWithDeviceCode(inProgress.userCodeInfo, err => {
-			// Callback is only called in failure scenarios.
-			if (err) {
-				console.warn(`Error while cancelling auto OAuth: ${err}`);
+	private async _initialize(storedAccounts: azdata.Account[]): Promise<azdata.Account[]> {
+		const accounts: azdata.Account[] = [];
+		for (let account of storedAccounts) {
+			const azureAuth = this.getAuthMethod(account);
+			if (!azureAuth) {
+				account.isStale = true;
+				accounts.push(account);
+			} else {
+				accounts.push(await azureAuth.refreshAccess(account));
 			}
-		});
-
-		return Promise.resolve();
+		}
+		this.initComplete.resolve();
+		return accounts;
 	}
 
-	private async clearAccountTokens(accountKey: azdata.AccountKey): Promise<void> {
-		// Put together a query to look up any tokens associated with the account key
-		let query = <adal.TokenResponse>{ userId: accountKey.accountId };
 
-		// 1) Look up the tokens associated with the query
-		// 2) Remove them
-		let results = await this._tokenCache.findThenable(query);
-		this._tokenCache.removeThenable(results);
+	getSecurityToken(account: azdata.Account, resource: azdata.AzureResource): Thenable<{}> {
```

Comment:
Shouldn't the type here be TokenResponse (and really should be TokenResponse | undefined given what _getSecurityToken can actually return)

Reply:
Yeah technically this is only called as an interface method defined 

https://github.com/microsoft/azuredatastudio/blob/43d9ed04746b736f35923a005e00dc3b81d72c62/src/sql/azdata.d.ts#L2277

But yeah since that accepts any object I'll do that and actually update the azdata definition as well.